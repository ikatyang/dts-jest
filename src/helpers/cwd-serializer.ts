export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
