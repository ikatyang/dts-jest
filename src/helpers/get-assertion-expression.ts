import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
