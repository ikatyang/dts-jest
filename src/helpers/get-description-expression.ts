import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
