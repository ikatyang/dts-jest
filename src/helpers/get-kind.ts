import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
