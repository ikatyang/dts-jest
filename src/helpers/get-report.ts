import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
