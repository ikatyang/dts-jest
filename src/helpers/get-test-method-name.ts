import {TriggerKind} from '../definitions';

export const get_test_method_name = (kind: TriggerKind) => {
  switch (kind) {
    case TriggerKind.None:
    case TriggerKind.Show:
      return 'test';
    case TriggerKind.Only:
      return 'test.only';
    case TriggerKind.Skip:
      return 'test.skip';
    default:
      throw new Error(`Unexpected kind ${JSON.stringify(kind)}`);
  }
};
