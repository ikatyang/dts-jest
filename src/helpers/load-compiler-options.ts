import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import {ITSConfig} from '../definitions';

export const load_compiler_options = (tsconfig: ITSConfig, cwd: string = process.cwd()): ts.CompilerOptions => {
  const {
    compilerOptions: compiler_options = {},
  } = tsconfig;

  if (typeof tsconfig.extends === 'string') {
    const filename = path.resolve(cwd, tsconfig.extends);
    const {
      extends: target_extends,
      compilerOptions: parent_compiler_options = {},
    } = JSON.parse(fs.readFileSync(filename, 'utf-8'));

    return load_compiler_options(
      {
        extends: target_extends,
        compilerOptions: {
          ...parent_compiler_options,
          ...compiler_options,
        },
      },
      path.dirname(filename),
    );
  }

  return compiler_options;
};
