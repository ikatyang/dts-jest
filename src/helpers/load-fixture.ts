import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
