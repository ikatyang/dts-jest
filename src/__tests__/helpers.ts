import * as path from 'path';
import * as ts from 'typescript';
import {create_snapshots} from '../helpers';

test('create_snapshots() should return correctly', () => {
  const filename = path.resolve(__dirname, '../../fixtures/snapshots/example.ts');
  const lines = [0, 3];
  const program = ts.createProgram([filename], {});
  expect(create_snapshots(program, filename, lines)).toMatchSnapshot();
});
