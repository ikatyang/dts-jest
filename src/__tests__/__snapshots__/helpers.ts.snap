// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create_snapshots() should return correctly 1`] = `
Object {
  "0": "string",
  "3": "Argument of type '\\"str\\"' is not assignable to parameter of type 'number'.",
}
`;
